src/asciidoctor-kroki.js

```diff
@@ -127,12 +127,13 @@ function diagramBlockMacro (name, context) {
     self.named(name)
     self.process((parent, target, attrs) => {
       let vfs = context.vfs
+      target = parent.applySubstitutions(target, ['attributes'])
       if (typeof vfs === 'undefined' || typeof vfs.read !== 'function') {
         vfs = require('./node-fs')
+        target = parent.getDocument().normalizeSystemPath(target)
```

Comment:
I believe that `normalizeSystemPath` is "safe" to call on the browser. We will need to add a test to make sure. I'm asking Dan to know if it's safe to call this method on an Antora resource identifier.
So in theory, we should be able to move this call outside of the `if` condition.

Reply:
I think `file://` and `http/s` paths have to be absolute (by definition).

But I think for the `file://` case we need to make sure the jail / safe mode criteria are respected.
So I think it is correct to call `normalizeSystemPath` for this case. It looks like the Asciidoctor PathResolver class detects and handles the different URI types.

I am a little skeptical it is possible for this method to be safe on an Antora resource id unless there's something I'm missing in the code. I wonder whether that means a carefully crafted relatively reference in an Antora xref can bypass Asciidoctor's safe mode?

Reply:
I think perhaps safe mode doesn't adversely affect Antora because the content aggregation phase doesn't allow content outside of well defined components to be included and so the need to consider jails doesn't arise because only allowed content is included in the vfs to begin with.

Reply:
From Dan:

> no, we should never communicate with the filesystem directly. only the content aggregator / classifier is permitted to do that.
(or whatever else is populating the content catalog)
if the file comes from the worktree, it's okay to read file.abspath (because then we know there is a direct correlation to the filesystem, but that doesn't need to be modified in any way)

So the test should be: normalize unless we are running in an Antora context.
We assume that we are running in an Antora context if `context.contentCatalog` is defined: https://github.com/Mogztter/asciidoctor-kroki/blob/912802169b3a7d6d1c3b020f81c2d1cb1916884b/src/asciidoctor-kroki.js#L150